consumption.tex

```diff
@@ -0,0 +1,62 @@
+We now shift our attention to the problem of verifying whether a block really belongs in the best source chain. We provide multiple ways of implementing the aforementioned \textsf{is-in-best-chain} method.
+
+\noindent
+\textbf{NIPoPoWs.}
+Verifying block connection can be achieved through NIPoPoWs, as in~\cite{pow-sidechains}.
+We remark that with this setup a block connection proof may be considered valid provisionally, but there needs to be a period in which the proof can be disputed for the smart contract to be certain for the validity of the proof. Specifically, when a user performs a claim they have to put down some collateral. If they have provided a valid NIPoPoW, \textsf{is-in-best-chain} will initially be true. However a challenger can dispute the provided proof which would turn the result of \textsf{is-in-best-chain} to false, abort the claim and grant the challenger the user's collateral. If the contestation period ends with the proof undisputed then the collateral gets returned to the user and the claim is performed successfully.
+
+\noindent
+\textbf{Federation.}
+A simpler approach is to allow a federation of nodes monitoring the source chain to vote for the blocks that are included in the best source chain.
+
+\import{./}{algorithms/alg.is-in-best-chain-federation.tex}
+
+Let us assume that the smart contract holds in its state the continuously updated set of all valid blocks on the best source blockchain, in the form of a Merkle Tree root $\mathcal{M}$. Then, by verifying a Merkle Tree inclusion proof \textsf{blockConnection} for $\textsf{b} \in \mathcal{M}$ a verifier can be certain that the block provided is indeed part of the best source blockchain. Combined with verifying the transaction this completes our {\sf is-in-best-chain}$_\mathcal{M}$ function, shown in Algorithm~\ref{alg.is-in-best-chain-federation}.
+
+We now address the issue of maintaining an up-to-date set of blocks in the form of $\mathcal{M}$. We propose a federation mechanism of source blockchain full nodes that monitor the source chain and vote for the new set of source chain blocks, if and when it changes. A majority of votes is necessary for $\mathcal{M}$ to be updated on the target blockchain. In this manner, the federation acts as a bridge from the source to the target blockchain. The security of our scheme relies on the assumption that the federation majority is honest.
+
+We propose encoding the honestly adopted chain by using Merkle Mountain Ranges (MMRs)~\cite{flyclient}. Specifically, to generate a checkpoint we take all the chain's block ids in sequence and use them as leaves in a Merkle Mountain Range. We then use the root of that MMR as our encoding. Our rationale for using MMRs instead of other kinds of Merkle Trees is that we expect the most frequent modifications to the tree to be appends (when the source chain is extended by one block), which is a very efficient operation to perform on an existing MMR.
```

Comment:
most frequent? not always?

Reply:
you could have reorgs

Reply:
we should state that federation members only sign stable blocks

Reply:
this applies to all mechanisms